src/system/random/basic.lean

```diff
@@ -0,0 +1,467 @@
+
+end stream
+
+variables {g : Type} [random_gen g]
+
+/-- generate a random bit vector of length `n` -/
+def bitvec.random (n : ℕ) : rand_g g (bitvec n) := do
+bs ← rand.random_series bool,
+pure ⟨bs.take n, bs.length_take _⟩
+
+section fit_in_Icc
+
+parameters {i' n : ℕ}
+parameters {x y : bitvec n}
+
+parameters P' : x.to_nat ≤ y.to_nat
+include P'
+
+local infix ^ := nat.pow
+
+lemma bitvec.interval_fits_in_word_size :
+  x.to_nat + i' % (1 + (y.to_nat - x.to_nat)) < 2^n :=
+begin
+  let x' := x.to_nat,
+  let y' := y.to_nat,
+  apply @lt_of_lt_of_le _ _ _ (x' + (y' - x' + 1)),
+  { apply add_lt_add_left, simp only [add_comm _ 1],
+    apply nat.mod_lt,
+    simp [add_comm 1,zero_lt_succ], },
+  { rw [← add_assoc,← nat.add_sub_assoc P',nat.add_sub_cancel_left,add_one],
+    clear P' i',
+    cases y with y Hy,
+    unfold bitvec.to_nat vector.to_list subtype.val bitvec.bits_to_nat,
+    rw [← reverse_reverse y, foldl_reverse,← Hy,← length_reverse],
+    generalize : reverse y = z,
+    clear x' x y' Hy y n,
+    induction z with x xs,
+    { rw [list.length,list.foldr,nat.pow] },
+    { simp [foldr,length,one_add,pow_succ,flip,bitvec.add_lsb,add_comm _ 1],
+      transitivity succ
+       (foldr (λ (b : bool) (a : ℕ), a + a + cond b 1 0) 0 xs +
+          foldr (λ (b : bool) (a : ℕ), a + a + cond b 1 0) 0 xs + 1),
+      { apply succ_le_succ, apply add_le_add_left,
+        cases x, apply nat.zero_le, refl, },
+      { simp! only,
+        rw [← nat.succ_eq_add_one,← nat.succ_add,← nat.add_succ],
+        rw [mul_comm _ 2,← two_mul],
+        apply nat.mul_le_mul_left,
+        simpa [flip,bitvec.add_lsb] using z_ih, } }, },
+end
+end fit_in_Icc
+
+open nat
+
+/-- Use `i` to generate an element of the interval `x .. y` -/
+protected def bitvec.fit_in_Icc {n : ℕ} (x y : bitvec n) (P : x ≤ y)
+  (i : bitvec n) :
+  (x .. y) :=
+let x' := x.to_nat,
+    y' := y.to_nat,
+    i' := i.to_nat,
+    r := i' % (y' - x' + 1) + x' in
+have Hx : x ≤ bitvec.of_nat n r,
+  begin
+    dsimp [r],
+    simp only [bitvec.le_def,bitvec.to_nat_of_nat,add_comm _ x',add_comm _ 1],
+    rw [mod_eq_of_lt],
+    { apply nat.le_add_right },
+    dsimp [x', y', i'],
+    apply bitvec.interval_fits_in_word_size,
+    apply P
+  end,
+have Hy : bitvec.of_nat n r ≤ y,
+  begin
+    dsimp [r],
+    rw [bitvec.le_def,bitvec.to_nat_of_nat,mod_eq_of_lt],
+    transitivity (y' - x') + x',
+    { apply add_le_add_right,
+      apply le_of_lt_succ,
+      rw ← add_one,
+      apply mod_lt,
+      rw add_one, apply zero_lt_succ },
+    { transitivity x' + (y' - x'),
+      apply le_of_eq, ac_refl,
+      rw [← nat.add_sub_assoc P,nat.add_sub_cancel_left], },
+    simp only [add_comm _ x',add_comm _ 1],
+    apply bitvec.interval_fits_in_word_size P,
+  end,
+⟨ bitvec.of_nat _ r , Hx , Hy ⟩
+
+instance random_bitvec (n : ℕ) : random (bitvec n) :=
+{ random := λ _ inst, @bitvec.random _ inst n,
+  random_r := λ _ inst x y p, bitvec.fit_in_Icc _ _ p <$> @bitvec.random _ inst n }
```

Comment:
`random_r` is not uniformly distributed if `y - x + 1` does not divide `2^n` (because the output of `bitvec.fit_in_Icc` isn't). I'd argue that the output of `random_r` should be uniformly distributed (or at least as close as we can get with a PRNG).

Reply:
We could do rejection sampling much like Haskell seems to do: https://hackage.haskell.org/package/random-1.2.0/docs/src/System.Random.Internal.html#unsignedBitmaskWithRejectionM.

That said, I don't quite see how the concrete sampling method is related to the random testing framework this PR is about. To me at least, a better sampling method is a completely independent PR. And also, PRNGs are a big can of worms where you really want to do proper testing and take your time because they are easy to get wrong in subtle ways. 

I'd really prefer if alternative sampling methods were done after this PR is in (and quite possibly by a different person for that matter, there is little reason why these two should be tied into one PR and block community collaboration).